dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * Assuming a structure like:
+ * <pre>
+ *
+ * TrackerBundleParams
+ *   |
+ *   __TEI
+ *      |_ENROLLMENT 1
+ *      |      |
+ *      |      |_ EVENT 1
+ *      |      |
+ *      |      |_ EVENT 2
+ *      |
+ *      |_ENROLLMENT 2
+ *            |
+ *            |_ EVENT 3
+ *            |_ EVENT 4
+ * </pre>
+ *
+ * This converter will transform the object into:
+ *
+ * <pre>
+ *
+ * TrackerBundleParams
+ *   |
+ *   __TEI
+ *  |__ENROLLMENT 1, ENROLLMENT 2
+ *  |
+ *  |__EVENT 1, EVENT 2, EVENT 3, EVENT 4
+ *
+ * </pre>
+ *
+ *
+ * @author Luciano Fiandesio
+ */
+public class TrackerBundleParamsConverter
+    extends
+    StdConverter<TrackerBundleParams, TrackerBundleParams>
+{
+    @Override
+    public TrackerBundleParams convert( TrackerBundleParams bundle )
+    {
+        if ( hasNestedStructure( bundle ) )
+        {
+            // collect a list of TEI ids
+            List<String> teiIds = bundle.getTrackedEntities().stream().map( TrackedEntity::getTrackedEntity )
+                .collect( Collectors.toList() );
+            
+            // collect all the enrollments of the TEIs
+            List<Enrollment> enrollments = bundle.getTrackedEntities().stream()
+                .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() );
+
+            for ( Enrollment enrollment : enrollments )
+            {
+                if ( teiIds.contains( enrollment.getTrackedEntityInstance() ) )
+                {
+                    bundle.getEnrollments().add( enrollment );
+                }
+                else
+                {
+                    if ( bundle.getAtomicMode().equals( AtomicMode.ALL ) )
+                    {
+                        fail( "Enrollment", enrollment.getEnrollment(), "Tracked Entity Instance" );
+                    }
+                }
+            }
+            List<String> enrollmentIds = bundle.getEnrollments().stream().map( Enrollment::getEnrollment )
+                .collect( Collectors.toList() );
+
+            List<Event> events = bundle.getEnrollments().stream().flatMap( l -> l.getEvents().stream() )
+                .collect( Collectors.toList() );
+            for ( Event event : events )
+            {
+                if ( enrollmentIds.contains( event.getEnrollment() ) )
+                {
+                    bundle.getEvents().add( event );
+                }
+                else
+                {
+                    if ( bundle.getAtomicMode().equals( AtomicMode.ALL ) )
+                    {
+                        fail( "Event", event.getEvent(), "Enrollment" );
+                    }
+                }
+            }
+            // remove the nested structure
+            for ( TrackedEntity tei : bundle.getTrackedEntities() )
+            {
+                tei.setEnrollments( Collections.emptyList()  );
+            }
+        }
+
+        return bundle;
+    }
+
+    private boolean hasNestedStructure( TrackerBundleParams bundle )
```

Comment:
@mortenoh not sure this is the best way to "decide" if a `TrackerBundleParams` should be flattened. Thoughts?

Reply:
Hm, i think this is fine.. the only thing this won't support i guess is it we also want top level enrollments which also includes events, not sure how common it is to be honest.. but it was supported before